### ISSUE TITLE
Distributed work failing connection errors build out completely

### ISSUE THREAD

**Issue by g3cko** [fastbuild/fastbuild#721]
With .99 I am seeing remote connection failures seemingly causing the build to fully exit instead of just flagging remote workers that fail to respond.

```
                // free the network distribution system (if there is one)
                FDELETE m_Client;
                m_Client = nullptr;
```
Does this sound accurate or intended?  

Even in a fully working office, we occasionally get machines going down but now with quarantine/work from home, this is extremely frequent to get a machine goes down and leaves it's "token" on the worker cache. 

Similarly, trying to get our tools to cleanly shutdown and give FBWorker a chance to clean this file up, I found WM_CLOSE is not handled so I wasn't able to externally (using taskkill /PID ...  but *without* /F) tell the worker to shutdown.

Adding this to OSWindow.cpp got me it to exit and cleanup.

```
            case WM_CLOSE:
            case WM_QUIT:
            {
                if ( window->OnQuit() )
                {
                    return 0; // Handled
                }
                break;
```

**Comment by ffulin** [comment]
*> With .99 I am seeing remote connection failures seemingly causing the build to fully exit instead of just flagging remote workers that fail to respond. Does this sound accurate or intended?*

This is not normal.

When an error occurs remotely, whether it causes the build to fail or not depends on the nature of the error. FASTBuild distinguishes compile errors (which cause the build to fail/stop) from "system errors" like a remote host being shutdown or out of disk space (which are effectively ignored, with the task being retried on another node).

You can look at ObjectNode::HandleSystemFailures if you want to see the various cases that are considered system failures (and will be retried by the host)

If you are seeing a remote failure that is not considered a "system error" but should be, and is causing the build to terminate when it shouldn't, it's possible there is a case that is not being handled correctly. If you can provide more information about the nature of the failure, I might be able to provide more information.

*> machine goes down and leaves it's "token" on the worker cache.*

This is perfectly normal and is handled correctly. If they worker is not available, FASTBuild will simply connect to another one. In the next version (v1.00) workers will cooperate to clean the brokerage of stale tokens as an optimization, but this isn't necessary for things to work.

*> I found WM_CLOSE is not handled*

It seems reasonable to extend the message pump to handle this case.

**Comment by g3cko** [comment]
Thanks @ffulin for your notes.   I hadn't fully gotten to the bottom of this so some speculation in the above.  

I'll check out ObjectNode::HandleSystemFailures and double check everything there again.  

And... I did just find some issues with missing extra file dlls that came with the latest VS 2019 update, so this may have been just that.  I was thrown off by the lack of any output errors to indicate what wasn't working, however, it seems that it was failing at startup so there wasn't any error output to report.  

This particular error is a bit hard to diagnose.  I'm not sure what the best way to improve it is though...  Ideally, if you could sandbox and hit this error locally... that would up there I think.

**Comment by ffulin** [comment]
When the compiler requires a DLL that is missing, if the DLL is implicitly loaded by the OS, then we get a nice return code (0xC000007B) (which FASTBuild detects) and prints an extra info about what to do.

Unfortunately, the recent addition to VS2019 (tbbmalloc.dll) is loaded by the compiler manually via LoadLibrary() and if it fails, the compiler exits in a different way (if I remember correctly with no useful info at all).

The way i figured out what the problem was, was to look at the list of dlls from the previous version, add any dlls I wasn't syncing and removed them one by one until I narrowed down the one I actually needed.

An alternative way might be to spin up a VM, copy the dlls into a folder and try starting the process. You could narrow it down the same way or use ProcessMonitor to figure out what it's trying to load.

In case you didn't see it, the documentation for Compiler() lists the required dlls: https://fastbuild.org/docs/functions/compiler.html and should be current

**Comment by g3cko** [comment]
The dll that caused this was vcruntime140_1.dll, which is pretty new.  I think it was the last VS19 update, but I'm not 100% on that.  This looks to be implicitly loaded though, if I'm understanding the dependencies correctly... 

And now that you mention it, it does return the error code which I didn't catch earlier. I didn't realize this wasn't an internal code.  Of course, after a quick google:  

> #define STATUS_DLL_NOT_FOUND             ((NTSTATUS)0xC0000135L)

And for reference, this is all that was in the output. 

`
PROBLEM: d:\...\some.obj
Failed to build Object. Error: 3221225781 (0xc0000135) Target: 'C:\Users\...\AppData\Local\Temp\.fbuild.tmp\0x00000000\core_1001\some.obj'
`

In addition to being installed to the VC redist folder, this is also in the system32 which explains how this "sometimes" worked.  

Since this was the second time I hit this type of error, I wanted to at least post about it.  Now I am thinking this specific case is handled by adding a case to HandleSystemFailures for that error code.  It's also good to know that error code is not an internal one.

**Comment by ffulin** [comment]
Interestingly I'm not syncing that dll and have remote compilation working, even on a box that has never had anything installed on it. I am using VS2019 v14.22.27905. I have seen some situations where certain DLLs are only with certain language features in use or other options, though this being the VC runtime that seems unlikely to be a factor.

In any case, the DLL is tiny so there's no real reason to worry about it - simply add it to your list.

It does look like 0xC0000135 should be added to the same check as 0xC000007B, where it prints a hint to check the Compiler settings.  (The case is explicitly **not** considered a "system failure" because it's a configuration problem you'd want to always know about). Looking at extra message that is emitted, it could probably be a lit bit more explicit about checking the .ExtraFiles to ensure all required DLLs are specified.

**Comment by Domain** [comment]
I have the same problem when upgrade VS2019 to 16.5.1 from 16.4.3
Add vcruntime140_1.dll to ExtraFiles solve it.

**Comment by ffulin** [comment]
I've confirmed that this new DLL is required by the latest version (16.5.1) and was not required by 16.4.4.

I've updated:
 - FASTBuild to be built with the latest version (including the DLL update in the Compiler() settings)
 - The remote error detection to check for 0xC0000135 and report extended erro information (the hint to check Compiler() settings)
 - The documentation for Compiler() to reference the new DLL in the examples

These changes (https://github.com/fastbuild/fastbuild/commit/abcccbbf793b12f9ec8b5dd012ec3871c7bb2a5b) will all be in the next release (v1.00)

**Comment by ffulin** [comment]
I think we can probably consider this issue resolved. As such, i'll mark it as "Resolved in Next Version" and close it when v1.00 is released.

If I've missed anything though, please let me know.

**Comment by g3cko** [comment]
Ah - yes, sorry forgot to come back to 'close' this but that sounds good.

Thanks!
